packages/lesswrong/components/comments/DebateTypingIndicator.tsx

```diff
@@ -0,0 +1,72 @@
+// TODO: Import component in components.ts
+import React, {useState} from 'react';
+import { fragmentTextForQuery, registerComponent } from '../../lib/vulcan-lib';
+import {useOnNotificationsChanged} from '../hooks/useUnreadNotifications';
+import {useCurrentUser} from '../common/withUser';
+import {TypingIndicatorMessage} from '../../client/serverSentEventsClient';
+import {useGlobalKeydown} from '../common/withGlobalKeydown';
+import {gql, useMutation} from '@apollo/client';
+import throttle from 'lodash/throttle';
+
+const styles = (theme: ThemeType): JssStyles => ({
+  root: {
+    ...theme.typography.commentStyle
+  },
+});
+
+export const DebateTypingIndicator = ({classes, post}: {
+  classes: ClassesType,
+  post: PostsWithNavigation | PostsWithNavigationAndRevision,
+}) => {
+
+  const [typingIndicators, setTypingIndicators] = useState<TypingIndicatorInfo[]>([]);
+  const currentUser = useCurrentUser();
+
+  const [upsertTypingIndicator] = useMutation(gql`
+    mutation upsertUserTypingIndicator($documentId: String!) {
+      upsertUserTypingIndicator(documentId: $documentId) {
+        ...TypingIndicatorsDefaultFragment
+      }
+    }
+    ${fragmentTextForQuery("TypingIndicatorsDefaultFragment")}
+  `)
+
+  useGlobalKeydown(throttle(() => {
+    void upsertTypingIndicator({variables: {documentId: post._id}})
+  }, 300));
+
+  useOnNotificationsChanged(currentUser, (messageString) => {
+    const message : TypingIndicatorMessage = JSON.parse(messageString)
```

Comment:
note: directly assigning the output of `JSON.parse` to something isn't super safe; generally you'd want to validate that it's actually the thing you're expecting.  In this case I think it's actually a `ConnectionMessage` rather than a `TypingIndicatorMessage`, and you need to check what the `eventType` value is before using it?  (And if it's a `NotificationCheckMessage`, you don't actually want to execute any logic on it in this component.)